[GLUTEN-3620][VL] Support Range operator for Velox Backend

ArnavBalyan opened the pull request:
## What changes were proposed in this pull request?
- Currently the range operator fallsback leads to R2C overhead for downstream operators in velox.
- Added support for RangeExec which produces range in columnar batches.
- In the future the range generation can itself be offloaded to velox backend. 

## How was this patch tested?
<img width="999" alt="image" src="https://github.com/user-attachments/assets/b32d5db1-a3fe-4647-9dc4-ef3bfdb0f2dd">

 - Plan reflects the new operator
 - Added unit tests

CodenameGHOST007 reviewed (commented):
Does velox currently have any similar operator that we can leverage or a new implementation will be needed in case we want to offload it.

ArnavBalyan commented:
> Does velox currently have any similar operator that we can leverage or a new implementation will be needed in case we want to offload it.

Yes, I'm planning to work on the offload implementation. Currently it seems unlikely we will have to introduce code in velox. Will raise PR once possible thanks. 
Some minor changes are needed for the tests, will enable them thanks for review

zhztheplayer commented on gluten-substrait/src/main/scala/org/apache/gluten/execution/RangeExecBaseTransformer.scala:
```diff
@@ -0,0 +1,85 @@
+    outputAttributes
+  }
+
+  override protected def doValidateInternal(): ValidationResult = {
+    val isSupported = BackendsApiManager.getSettings.supportRangeExec()
+
+    if (!isSupported) {
+      return ValidationResult.failed(
+        s"RangeExec is not supported by the current backend."
+      )
+    }
+    ValidationResult.succeeded
+  }
```
I am opening https://github.com/apache/incubator-gluten/pull/8177 that may simplify the usage of validators. Perhaps could do a rebase once that PR merged.

ArnavBalyan replied:
Sure, rebased since #8177 is merged, I believe we won't need to add to [this validator](https://github.com/apache/incubator-gluten/blob/36f92a529450b247903ef57d07b7f243d31ee302/gluten-substrait/src/main/scala/org/apache/gluten/extension/columnar/validator/Validators.scala#L225) since currently the code does not interact with native or rely on native validation. It may be needed once offloaded to velox

zhztheplayer replied:
@ArnavBalyan The validator's name is a little bit misleading, though I think it's the one to call `doValidate`, then `doValidateInternal` to make sure `BackendsApiManager.getSettings.supportRangeExec()` is checked. Would you like to double check? Or is there another code path calling `RangeExecTransformer.doValidateInternal`?

ArnavBalyan replied:
Yes makes sense, let me double check and get back

ArnavBalyan replied:
Just double checked, the this is defined in offloadothers, which is wired upto validatorBuilder that calls doValidate and then doValidateInternal. This will trigger the check, UTs also confirm it, I see it's currently similarly implemented for other columnar java operators as well thanks!

zhztheplayer replied:
> Just double checked, the this is defined in offloadothers, which is wired upto validatorBuilder that calls doValidate and then doValidateInternal. This will trigger the check, UTs also confirm it, I see it's currently similarly implemented for other columnar java operators as well thanks!

Got it. Thanks for the double-checking. I forgot to mention that after https://github.com/apache/incubator-gluten/pull/8177 we don't have to add case lines to that validator. The new code should call the native validation API automatically. Thanks!

zhouyuan commented:
> > Does velox currently have any similar operator that we can leverage or a new implementation will be needed in case we want to offload it.
> 
> Yes, I'm planning to work on the offload implementation. Currently it seems unlikely we will have to introduce code in velox. Will raise PR once possible thanks. Some minor changes are needed for the tests, will enable them thanks for review

We have some early discussion on this to map it to `unnest + sequence ` - but didn't try the real impl at that time
#3620

zhouyuan commented:
@ArnavBalyan would you please check the failed unit tests? 

thanks, -yuan

ArnavBalyan commented:
Hi @zhouyuan thanks for the review, interesting thread, I will go through! I'll fix the tests today/tomorrow thanks!

ArnavBalyan commented on gluten-ut/spark33/src/test/scala/org/apache/spark/sql/execution/GlutenSQLRangeExecSuite.scala:
```diff
@@ -0,0 +1,94 @@
+import org.apache.spark.sql.GlutenSQLTestsTrait
+import org.apache.spark.sql.Row
+
+class GlutenSQLRangeExecSuite extends GlutenSQLTestsTrait {
```
Enabled

ArnavBalyan commented on backends-velox/src/main/scala/org/apache/gluten/execution/RangeExecTransformer.scala:
```diff
@@ -0,0 +1,139 @@
+ * @param child
+ *   Child SparkPlan nodes for this operator, if any.
+ */
+case class RangeExecTransformer(
```
Updated this thanks, I am planning to add native support soon, will name it accordingly once it's done

ArnavBalyan commented on backends-velox/src/main/scala/org/apache/gluten/execution/RangeExecTransformer.scala:
```diff
@@ -0,0 +1,139 @@
+                current += numRows * step
+
+                val batch = new ColumnarBatch(vectors.asInstanceOf[Array[ColumnVector]], numRows)
+                val offloadedBatch = ColumnarBatches.offload(allocator, batch)
```
Thanks for the feedback this is much more elegant, updated it!
![image](https://github.com/user-attachments/assets/7404a727-0260-41a0-a945-5aef6e322be2)

ArnavBalyan commented on gluten-ut/spark32/src/test/scala/org/apache/spark/sql/execution/GlutenSQLRangeExecSuite.scala:
```diff
@@ -0,0 +1,94 @@
+import org.apache.spark.sql.GlutenSQLTestsTrait
+import org.apache.spark.sql.Row
+
+class GlutenSQLRangeExecSuite extends GlutenSQLTestsTrait {
```
Updated to keep it only under 33

ArnavBalyan commented:
The unit tests are failing due to another bug, this PR will be rebased once the dependent PR is merged. Which will also fix the UTs.
https://github.com/apache/incubator-gluten/issues/8453

ArnavBalyan commented:
Hi @zhztheplayer had addressed the comments can you pls take a look. Also it seems some UTs in CSV suite are failing and seem flaky. It does not seem related to this change, just wanted to confirm if it's expected thanks

zhouyuan commented:
@ArnavBalyan 
The code looks in good shape now, would you please also check the CK backend failure?
I think you may need to write a dummy impl for CK

```
18:16:15  [ERROR] /home/jenkins/agent/workspace/gluten/gluten-ci/ut-stage-5/backends-clickhouse/src/main/scala/org/apache/gluten/backendsapi/clickhouse/CHSparkPlanExecApi.scala:67: error: class CHSparkPlanExecApi needs to be abstract.
18:16:15  [ERROR] Missing implementation for member of trait SparkPlanExecApi:
18:16:15  [ERROR]   def genColumnarRangeExec(start: Long, end: Long, step: Long, numSlices: Int, numElements: BigInt, outputAttributes: Seq[org.apache.spark.sql.catalyst.expressions.Attribute], child: Seq[org.apache.spark.sql.execution.SparkPlan]): org.apache.gluten.execution.RangeExecBaseTransformer = ???
18:16:15  [ERROR] 
18:16:15  [ERROR] class CHSparkPlanExecApi extends SparkPlanExecApi with Logging {
18:16:15  [ERROR]       ^
18:16:15  [WARNING] 3 warnings
18:16:15  [ERROR] 1 error
18:16:15  [ERROR] exception compilation error occurred!!!
```

Cc: @taiyang-li as he is working on `RangeExec `for CK
https://github.com/apache/incubator-gluten/pull/8518

zhztheplayer reviewed (approved):
Overall looking good to me. Would you like to address the CH CI error @zhouyuan just mentioned? Thanks!

ArnavBalyan commented:
> Overall looking good to me. Would you like to address the CH CI error @zhouyuan just mentioned? Thanks!

Updated it

ArnavBalyan commented:
Hi @zhouyuan could you share the output of the latest ClickHouse CI, it seems something new popped up thanks

zhouyuan commented:
Hi @ArnavBalyan 
Here's the new failure:
```
[2025-01-16T08:22:49.210Z] - gluten-7077 bug in cross broad cast join *** FAILED ***
[2025-01-16T08:22:49.210Z]   org.apache.gluten.exception.GlutenNotSupportException: ColumnarRange is not supported in ch backend.
[2025-01-16T08:22:49.210Z]   at org.apache.gluten.backendsapi.clickhouse.CHSparkPlanExecApi.genColumnarRangeExec(CHSparkPlanExecApi.scala:943)
[2025-01-16T08:22:49.210Z]   at org.apache.gluten.extension.columnar.offload.OffloadOthers$ReplaceSingleNode.doReplace(OffloadSingleNodeRules.scala:336)
[2025-01-16T08:22:49.210Z]   at org.apache.gluten.extension.columnar.offload.OffloadOthers.offload(OffloadSingleNodeRules.scala:182)
[2025-01-16T08:22:49.210Z]   at org.apache.gluten.extension.columnar.heuristic.LegacyOffload$$anonfun$$nestedInanonfun$apply$1$1.applyOrElse(LegacyOffload.scala:28)
[2025-01-16T08:22:49.210Z]   at org.apache.gluten.extension.columnar.heuristic.LegacyOffload$$anonfun$$nestedInanonfun$apply$1$1.applyOrElse(LegacyOffload.scala:28)
[2025-01-16T08:22:49.210Z]   at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformUpWithPruning$2(TreeNode.scala:515)
[2025-01-16T08:22:49.210Z]   at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(origin.scala:76)
[2025-01-16T08:22:49.210Z]   at org.apache.spark.sql.catalyst.trees.TreeNode.transformUpWithPruning(TreeNode.scala:515)
[2025-01-16T08:22:49.210Z]   at org.apache.spark.sql.catalyst.trees.TreeNode.transformUp(TreeNode.scala:488)
[2025-01-16T08:22:49.210Z]   at org.apache.gluten.extension.columnar.heuristic.LegacyOffload.$anonfun$apply$1(LegacyOffload.scala:28)
```

For CK backend Jenkins log please check here:
https://github.com/apache/incubator-gluten/blob/main/docs/get-started/ClickHouse.md#new-ci-system

Thanks, -yuan

ArnavBalyan commented:
> Hi @ArnavBalyan Here's the new failure:
> 
> ```
> [2025-01-16T08:22:49.210Z] - gluten-7077 bug in cross broad cast join *** FAILED ***
> [2025-01-16T08:22:49.210Z]   org.apache.gluten.exception.GlutenNotSupportException: ColumnarRange is not supported in ch backend.
> [2025-01-16T08:22:49.210Z]   at org.apache.gluten.backendsapi.clickhouse.CHSparkPlanExecApi.genColumnarRangeExec(CHSparkPlanExecApi.scala:943)
> [2025-01-16T08:22:49.210Z]   at org.apache.gluten.extension.columnar.offload.OffloadOthers$ReplaceSingleNode.doReplace(OffloadSingleNodeRules.scala:336)
> [2025-01-16T08:22:49.210Z]   at org.apache.gluten.extension.columnar.offload.OffloadOthers.offload(OffloadSingleNodeRules.scala:182)
> [2025-01-16T08:22:49.210Z]   at org.apache.gluten.extension.columnar.heuristic.LegacyOffload$$anonfun$$nestedInanonfun$apply$1$1.applyOrElse(LegacyOffload.scala:28)
> [2025-01-16T08:22:49.210Z]   at org.apache.gluten.extension.columnar.heuristic.LegacyOffload$$anonfun$$nestedInanonfun$apply$1$1.applyOrElse(LegacyOffload.scala:28)
> [2025-01-16T08:22:49.210Z]   at org.apache.spark.sql.catalyst.trees.TreeNode.$anonfun$transformUpWithPruning$2(TreeNode.scala:515)
> [2025-01-16T08:22:49.210Z]   at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(origin.scala:76)
> [2025-01-16T08:22:49.210Z]   at org.apache.spark.sql.catalyst.trees.TreeNode.transformUpWithPruning(TreeNode.scala:515)
> [2025-01-16T08:22:49.210Z]   at org.apache.spark.sql.catalyst.trees.TreeNode.transformUp(TreeNode.scala:488)
> [2025-01-16T08:22:49.210Z]   at org.apache.gluten.extension.columnar.heuristic.LegacyOffload.$anonfun$apply$1(LegacyOffload.scala:28)
> ```
> 
> For CK backend Jenkins log please check here: https://github.com/apache/incubator-gluten/blob/main/docs/get-started/ClickHouse.md#new-ci-system
> 
> Thanks, -yuan

Thanks, this is super helpful, have addressed the UT failure

ArnavBalyan commented:
Hi @zhztheplayer @zhouyuan would it be possible to merge this, I was planning to rebase #8566 after it thanks!

zhztheplayer commented on gluten-substrait/src/main/scala/org/apache/gluten/execution/RangeExecBaseTransformer.scala:
```diff
@@ -0,0 +1,69 @@
+ * Base class for RangeExec transformation, can be implemented by the by supported backends.
+ * Currently velox is supported.
+ */
+abstract class RangeExecBaseTransformer(
```
Can we rename it to `ColumnarRangeBaseExec`? Given it's not a Substrait operator. 

It feels a similar issue is in PR https://github.com/apache/incubator-gluten/pull/8566 as well.

ArnavBalyan replied:
Updated both

zhztheplayer reviewed (commented):
Just a minor. Thank you for the contribution.

ArnavBalyan commented:
> Just a minor. Thank you for the contribution.

Updated this
